### PR TITLE
Improving CentralReport security and integration on operating systems

### PR DESCRIPTION
**Pull request by miniche**
This pull request will change many elements in CentralReport. Summary:
- [x] Adding a system user to secure CentralReport daemon.
- [x] Adding a launch script in /usr/local/bin, which will be included in the PATH.
- [x] All python modules will go to /usr/local/lib.

Now, with details:
## CentralReport user

For the moment, CentralReport is executed with root permissions. It isn't really secure, we can't say otherwise. So, to secure scripts, CentralReport will have his own user, with only few rights in the system.
- [x] Make some research
- [x] Add the new user on Debian
- [x] Add the new user on Mac OS
- [x] Update all install/uninstall functions
- [x] Update rights on /usr/local/lib/centralreport (see "Launch script")
- [x] Make a new folder in /var/log and update rights
- [x] Make a new folder in /var/run and update rights
- [x] Update documentation
## Launch script

For the moment, we must start or stop CentralReport with a long command: `sudo python /usr/local/bin/centralreport/centralreport.py start`. Yeah... Boring.
By adding a new script, it will be easier. The new command can look like: `centralreport start`. Better, no?
- [x] Creation of the new binary script (CentralReport user is required)
- [x] Update install and uninstall bash scripts for all OS (copy, delete and chmod +x)
- [x] Update init script (init.d for Debian and LaunchDaemon for Mac)
- [x] Update documentation
## Python modules

Put all python scripts and modules in /usr/local/lib is more logical according to me.
- [x] Update all bash functions (install/uninstall)
- [x] Update documentation

### CODE REVIEW

**Comment by miniche** [comment]
In the commit 00ac8ae, I made some changes:
- "centralreport" will be available in $PATH with the script in /usr/local/bin/centralreport
- All CentralReport libraries are now stored in /usr/local/lib/centralreport
- A "CentralReport user" is now created on the system for more security. It's a system account, with no passord: it can't connect on the system, and programs can only launched by root user.
- Logs are now saved in /var/log/centralreport/. This folder belongs to the "CentralReport user".
- PID file are now saved in /var/run/centralreport/. This folder belongs to the "CentralReport user" too.

Also, all 3nd party libraries are now shipped with CentralReport in the "cr/libs" folder. Why?
- The CentralReport installer doesn't require an Internet connection yet (may be useful later)
- CentralReport doesn't depend on installed python libraries anymore: if the administrator updates python libraries installed on the system, CR couldn't be broken with an incompatible version of a library for example.
- CentralReport remains silent and can have a minimal impact on the system with this solution.

:warning:  **Please don't try to install CR on any host!**
The install.sh and uninstall.sh aren't finished yet. I will work in this during this week-end.
And, because of the new architecture, a uninstall script of the 0.1.0 version should be created.

**Comment by miniche** [comment]
I forgot some useful links.
About modules in a zip archive:
- [PEP 273](http://www.python.org/dev/peps/pep-0273/)
- [zipimport python module](http://docs.python.org/2/library/zipimport.html) (included in Python 2.3+)
- [Add a folder/zip to the sys.path list](http://docs.python.org/2/library/sys.html#sys.path)
- [Stack overflow](http://stackoverflow.com/questions/2886850/how-to-import-a-zip-file-to-my-py) (how to)

About the security of daemons:
- [An excellent stack exchange answer](http://unix.stackexchange.com/questions/29159/why-is-it-recommended-to-create-a-group-and-user-for-some-applications)

**Comment by miniche** [comment]
New known issue after last commits: values configured in the configuration file are not parsed anymore.
I'm investigating!

[EDIT]
Fixed with commit 24b3bc2

**Comment by miniche** [comment]
And also, for the error message on Mac OS when stopping CentralReport with the bin script ("shell-init: error retrieving current directory: getcwd: cannot access parent directories: Permission denied"): this doesn't occur every time.
Perhaps an error on the centralreport.log file. I'm investigating too!

[EDIT]
Fixed with the work on binary script and python daemon (commits 5707e96, 9fb9042, 2fed8d6 and 15ebf75)

**Comment by miniche** [comment]
Now, CentralReport daemon is able to catch some posix signals: SIGTERM and SIGINT.

When the user wants to stop CR, the binary script will send the "SIGTERM" signal to the current daemon, with the PID number (stored in /var/run). CR daemon can catch this signal, terminate all threads and exit itself.
If CR has not exited after 3 seconds, the binary script will send the "SIGKILL" signal. This signal is not catchable by any process, and it forces the OS to kill immediately the process.

CentralReport can receive the SIGINT signal. In this case, CR will exit immediately too. This is useful when CentralReport is executed in a console or in debug mode.
:warning: **Debug mode not works in a console yet.**

For more informations about signals:
http://en.wikipedia.org/wiki/Unix_signal

**Comment by miniche** [comment]
Remains to be done in this branch:
- A script for transition between 0.1.0 and 0.2.0 (deleting old files)
- Merging mac and debian functions in bash scripts (on the model of the new binary script)
- A big review!

**Comment by Guimove** [comment]
For de migration script, we can use the actual uninstall.sh, he already remove all file forme the develop branch.

**Comment by miniche** [comment]
I think... A big review is waiting for you :smile:
I will do more tests on Mac/Debian and Ubuntu, with different versions. But all seems good for me now.

File: bash/010_uninstaller.inc.sh [hunk 1]
```diff
@@ -0,0 +1,69 @@
+#!/bin/sh
+
+# ------------------------------------------------------------
+# CentralReport Unix/Linux "0.1.0 uninstaller"
+# Alpha version. Don't use in production environment!
+# ------------------------------------------------------------
+# https://github.com/miniche/CentralReport/
+# ------------------------------------------------------------
+
+# This script contains functions used to remove 0.1.0 from the current host.
```
Comment (Guimove) [inline bash/010_uninstaller.inc.sh]: This script contains functions used to remove CentralReport 0.1.0 from the current host.

File: bash/functions.inc.sh [hunk 2]
```diff
@@ -1,79 +1,823 @@
...
+}
+
+
+# --
+# Functions related to installation
+# --
+
+function copy_bin(){
+
+    # In some cases, /usr/local and /usr/local/bin doesn't exist. We will creating them in this case.
```
Comment (Guimove) [inline bash/functions.inc.sh]: We will create them in this case.

File: bash/functions.inc.sh [hunk 3]
```diff
@@ -1,79 +1,823 @@
...
+            return ${RETURN_CODE}
+        fi
+    fi
+
+    return 0
+}
+
+function copy_lib(){
+
+    # In some cases, /usr/local and /usr/local/bin doesn't exist. We will creating them in this case.
```
Comment (Guimove) [inline bash/functions.inc.sh]: We will create..

File: bash/functions.inc.sh [hunk 4]
```diff
@@ -1,79 +1,823 @@
...
+
+            # Getting the last Unique ID available for system users.
+            # From: http://superuser.com/questions/553374/find-available-ids-lower-than-500-vis-dscl
+            local continue="no"
+            local number_used="dontknow"
+            local fnumber_work_backwards_from=499
+            local fnumber=${fnumber_work_backwards_from}
+            local user_id=0
+            until [ ${continue} = "yes" ] ; do
+                if [ `dscl . -list /Users UniqueID | awk '{print $2, "\t", $1}' | sort -ug | grep -c "${fnumber}"` -gt 0 ] ; then
```
Comment (Guimove) [inline bash/functions.inc.sh]: More than 120 letters, you'll be whipped for that

File: bash/functions.inc.sh [hunk 5]
```diff
@@ -1,79 +1,823 @@
...
+                    logFile "Creating a new group for CentralReport..."
+
+                    # We get the last Primary Group ID available for system groups.
+                    continue="no"
+                    number_used="dontknow"
+                    fnumber_work_backwards_from=499
+                    fnumber=${fnumber_work_backwards_from}
+                    GROUP_UNIQUE_ID=0
+                    until [ ${continue} = "yes" ] ; do
+                        if [ `dscl . -list /Users UniqueID | awk '{print $2, "\t", $1}' | sort -ug | grep -c "${fnumber}"` -gt 0 ] ; then
```
Comment (Guimove) [inline bash/functions.inc.sh]: More than 120 letters, you'll be whipped for that

**Comment by miniche** [comment]
Ping @Ninir

**Comment by Guimove** [comment]
ping -t @Ninir

File: bash/functions.inc.sh (hunk 2 above)
Comment (Ninir) [inline bash/functions.inc.sh]: As you have 2 paths (`/usr/local` and `/usr/local/bin`) you should not use "doesn't" IMO.

File: bash/functions.inc.sh [hunk 6]
```diff
@@ -1,79 +1,823 @@
...
+    fi
+    if [ ! -d "/usr/local/lib" ]; then
+         execute_privileged_command mkdir /usr/local/lib
+    fi
+
+    execute_privileged_command mkdir ${CR_LIB_DIR}
+    RETURN_CODE="$?"
+
+    if [ ${RETURN_CODE} -ne "0" ]; then
+          logError "Error creating CentralReport library directory at ${CR_LIB_DIR} (Error code: ${RETURN_CODE})"
```
Comment (Ninir) [inline bash/functions.inc.sh]: `Error creating CentralReport the "libraries" directory...`

File: bash/functions.inc.sh [hunk 7]
```diff
@@ -1,79 +1,823 @@
...
+            logError "Error copying startup plist at ${STARTUP_PLIST} (Error code: ${RETURN_CODE})"
+            return ${RETURN_CODE}
+        fi
+
+    else
+        cp -f -v ${STARTUP_DEBIAN_INSTALL} ${STARTUP_DEBIAN}
+        RETURN_CODE="$?"
+
+        if [ ${RETURN_CODE} -ne "0" ]; then
+            logError "Error copying startup script at ${STARTUP_PLIST} (Error code: ${RETURN_CODE})"
```
Comment (Ninir) [inline bash/functions.inc.sh]: If we are creating our own startup script, then we should use `the` prefixing `startup` AFAIR
Ex: `Error copying the startup script`.

File: bash/functions.inc.sh [hunk 8]
```diff
@@ -1,79 +1,823 @@
...
+            return ${RETURN_CODE}
+        else
+            chmod 755 ${STARTUP_DEBIAN}
+
+            update-rc.d centralreport defaults
+            RETURN_CODE="$?"
+
+            if [ ${RETURN_CODE} -ne "0" ]; then
+                logError "Error registering startup script with update-rc.d (Error code: ${RETURN_CODE})"
```
Comment (Ninir) [inline bash/functions.inc.sh]: Same here

File: bash/functions.inc.sh [hunk 9]
```diff
@@ -1,79 +1,823 @@
...
+        fi
+    fi
+
+    return 0
+}
+
+function create_config_directory(){
+
+    if [ -d ${CR_CONFIG_DIR} ]; then
+        logFile "Config directory already exist!"
```
Comment (Ninir) [inline bash/functions.inc.sh]: `The config directory already exists`

File: bash/functions.inc.sh [hunk 10]
```diff
@@ -1,79 +1,823 @@
...
+        logError "Error updating owner of ${CR_CONFIG_DIR} (Error code: ${RETURN_CODE})"
+        return ${RETURN_CODE}
+    fi
+
+    return 0
+}
+
+function create_log_directory(){
+    if [ -d ${CR_LOG_DIR} ]; then
+        logFile "Log directory already exist!"
```
Comment (Ninir) [inline bash/functions.inc.sh]: `The log directory already exists!`

File: utils/init/debian/centralreport [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+# Default-Start:     2 3 4 5
+# Default-Stop:      0 1 6
+# Short-Description: Start CentralReport at boot time
+# Description:       Enable service provided by CentralReport Daemon.
+### END INIT INFO
+
+case $1 in
+    start | stop | restart | status)
+        /usr/local/bin/centralreport $*
+
```
Comment (Ninir) [inline utils/init/debian/centralreport]: Is this newline required?

File: utils/bin/centralreport [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+
+# ------------------------------------------------------------
+# CentralReport Unix/Linux - Binary script
+# Alpha version. Don't use in production environment!
+# ------------------------------------------------------------
+# https://github.com/miniche/CentralReport/
+# ------------------------------------------------------------
+
+# This script allows to manage the CentralReport daemon.
+# You can use one of the following argument:
```
Comment (Ninir) [inline utils/bin/centralreport]: I would add a `s` to arguments. Find it weird not to have it

File: utils/bin/centralreport [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+CR_DIR_LOG=/var/log/centralreport
+
+CR_USER_MAC=_centralreport
+CR_GROUP_MAC=_centralreport
+
+CR_USER_DEBIAN=centralreport
+CR_GROUP_DEBIAN=centralreport
+
+#
+# Executes bash commands, with administrative privileges.
```
Comment (Ninir) [inline utils/bin/centralreport]: In english you don't use commas this way (this one is not required for instance)

File: utils/bin/centralreport [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+#
+# PARAMETERS:
+#   $1..n: The command to execute.
+# RETURN: None
+#
+function execute_command() {
+    if [ $(uname -s) == "Darwin" ]; then
```
Comment (Ninir) [inline utils/bin/centralreport:36]: g1363779876588306776

File: bash/functions.inc.sh [hunk 15]
```diff
@@ -1,79 +1,823 @@
...
+
+        if [ -f ${STARTUP_PLIST} ]; then
+
+#            execute_privileged_command launchctl unload -w ${STARTUP_PLIST}
+#            RETURN_CODE="$?"
+#
+#            if [ ${RETURN_CODE} -ne "0" ]; then
+#                logError "Error unloading startup plist file with launchctl (Error code: ${RETURN_CODE})"
+#                return ${RETURN_CODE}
+#            fi
```
Comment (Ninir) [inline bash/functions.inc.sh]: If these lines are commented/not used anymore, you should remove them

File: bash/functions.inc.sh [hunk 16]
```diff
@@ -1,79 +1,823 @@
...
+
+            execute_privileged_command rm -f ${STARTUP_PLIST}
+            RETURN_CODE="$?"
+
+            if [ ${RETURN_CODE} -ne "0" ]; then
+                logError "Error deleting startup plist file at ${STARTUP_PLIST} (Error code: ${RETURN_CODE})"
```
Comment (Ninir) [inline bash/functions.inc.sh]: `Error deleting the startup plist...`

File: bash/functions.inc.sh [hunk 17]
```diff
@@ -1,79 +1,823 @@
...
+                return ${RETURN_CODE}
+            else
+                logFile "Startup plist removed"
+            fi
+        else
+            logInfo "Startup plist file not found!"
+        fi
+
+    else
+        logFile "Removing startup script..."
```
Comment (Ninir) [inline bash/functions.inc.sh]: `Removing the startup script...`

File: bash/functions.inc.sh [hunk 18]
```diff
@@ -1,79 +1,823 @@
...
+
+        if [ -f ${STARTUP_DEBIAN} ]; then
+            rm -rf ${STARTUP_DEBIAN}
+            RETURN_CODE="$?"
+
+            if [ ${RETURN_CODE} -ne "0" ]; then
+                logError "Error deleting startup script at ${STARTUP_DEBIAN} (Error code: ${RETURN_CODE})"
```
Comment (Ninir) [inline bash/functions.inc.sh]: `THE` comment

File: bash/functions.inc.sh [hunk 19]
```diff
@@ -1,79 +1,823 @@
...
+                return ${RETURN_CODE}
+            else
+
+                update-rc.d -f centralreport remove
+                RETURN_CODE="$?"
+
+                if [ ${RETURN_CODE} -ne "0" ]; then
+                    logError "Error removing startup script with update-rc.d (Error code: ${RETURN_CODE})"
```
Comment (Ninir) [inline bash/functions.inc.sh]: `THE` comment

File: bash/functions.inc.sh [hunk 20]
```diff
@@ -1,79 +1,823 @@
...
+
+function delete_log_directory(){
+    logFile "Removing log directory..."
+
+    if [ -d ${CR_LOG_DIR} ]; then
+        execute_privileged_command rm -R -f ${CR_LOG_DIR}
+        RETURN_CODE="$?"
+
+        if [ $? -ne "0" ]; then
+            logError "Error deleting log directory at ${CR_LOG_DIR} (Error code: ${RETURN_CODE})"
```
Comment (Ninir) [inline bash/functions.inc.sh]: `Error deleting the log directory`

File: bash/functions.inc.sh [hunk 21]
```diff
@@ -1,79 +1,823 @@
...
-    sh -c "$*" >> /dev/null 2>&1
-    local ret=$?
+function create_cr_user(){
 
-    if [ ${ret} -ne 0 ]; then
-        logFile "[ERR] ${message}"
-        echo -e "\r\033[0;31m [ERR]\033[0m ${message}"
+    RETURN_CODE=$(verify_cr_user)
+    if [ ${RETURN_CODE} -ne 0 ]; then
+        logFile "CentralReport user already exist. Skipping this step."
```
Comment (Ninir) [inline bash/functions.inc.sh]: `CentralReport user already exists`

File: bash/functions.inc.sh [hunk 22]
```diff
@@ -1,79 +1,823 @@
...
+                else
+                    user_id="${fnumber}"
+                    continue="yes"
+                fi
+            done;
+
+            logFile "New UniqueID available: ${user_id}"
+
+            if [ ${user_id} -eq 0 ]; then
+                logError "Enable to find an available UniqueID for CentralReport user."
```
Comment (Ninir) [inline bash/functions.inc.sh]: `Unable`

for (the) CentralReport user(s): `the` or `users`?

File: bash/functions.inc.sh [hunk 23]
```diff
@@ -1,79 +1,823 @@
...
+                return 1
+            elif [ ${user_id} -le 100 ]; then
+                logError "Enable to find an available UniqueID for CentralReport greater than 100."
```
Comment (Ninir) [inline bash/functions.inc.sh]: `Unable to find an available UniqueID, greater than 100, for the CentralReport user.

File: bash/functions.inc.sh [hunk 24]
```diff
@@ -1,79 +1,823 @@
...
+                logError "UID ${user_id} is already in use"
+                return 1
+            fi
+
+            # Checks if CentralReport group already exists.
+            GROUP_UNIQUE_ID=$(verify_cr_group)
+            if [ ${GROUP_UNIQUE_ID} -ne 0 ]; then
+                logFile "CentralReport group already exists. Skipping this step."
+            else
+
```
Comment (Ninir) [inline bash/functions.inc.sh]: Is this newline required?

File: bash/functions.inc.sh [hunk 25]
```diff
@@ -1,79 +1,823 @@
...
+                        else
+                            GROUP_UNIQUE_ID="${fnumber}"
+                            continue="yes"
+                        fi
+                    done;
+
+                    logFile "New PrimaryGroupID available: ${GROUP_UNIQUE_ID}"
+
+                    if [ ${user_id} -eq 0 ]; then
+                        logError "Enable to find an available PrimaryGroupID for CentralReport group."
```
Comment (Ninir) [inline bash/functions.inc.sh]: `Unable to find an available PrimaryGroupID for the CentralReport group.`

File: bash/functions.inc.sh [hunk 26]
```diff
@@ -1,79 +1,823 @@
...
+                        return 1
+                    elif [ ${user_id} -le 100 ]; then
+                        logError "Enable to find an available PrimaryGroupID for CentralReport group greater than 100."
```
Comment (Ninir) [inline bash/functions.inc.sh]: `Unable to find an available PrimaryGroupID for the CentralReport group greater than 100.`

File: bash/functions.inc.sh [hunk 27]
```diff
@@ -1,79 +1,823 @@
...
+
+                # Now, we must check in the UGID is really available.
+                if dscl . -readall /Groups | grep -q "PrimaryGroupID: *${GROUP_UNIQUE_ID}$" ; then
+                    logError "UGID ${GROUP_UNIQUE_ID} is already in use"
+                    return 1
+                fi
+
+                # Good, isn't it? Victoire de canard!
+                # We can add our group now!
+                logFile "Creating CentralReport group..."
```
Comment (Ninir) [inline bash/functions.inc.sh]: `Creating the CentralReport group...`

File: bash/functions.inc.sh [hunk 28]
```diff
@@ -1,79 +1,823 @@
...
+                    logFile "Error creating the CentralReport group."
+                    return 1
+                fi
+
+                logFile "Group successfully created!"
+            fi
+
+            logFile "Creating CentralReport user..."
+
+            # Now, we can create our user. The victory is near!
```
Comment (Ninir) [inline bash/functions.inc.sh]: `is close!`

File: bash/functions.inc.sh [hunk 29]
```diff
@@ -1,79 +1,823 @@
...
+            sudo dscl . -create /Users/_centralreport
+            sudo dscl . -create /Users/_centralreport UserShell /bin/bash
+            sudo dscl . -create /Users/_centralreport RealName "CentralReport daemon"
+            sudo dscl . -create /Users/_centralreport UniqueID ${user_id}
+            sudo dscl . -create /Users/_centralreport PrimaryGroupID ${GROUP_UNIQUE_ID}
+            sudo dscl . -create /Users/_centralreport NFSHomeDirectory /usr/local/lib/centralreport
+
+            # Hidding the user
```
Comment (Ninir) [inline bash/functions.inc.sh]: Hiding

File: bash/functions.inc.sh [hunk 30]
```diff
@@ -1,79 +1,823 @@
...
+            # http://superuser.com/questions/70156/hide-users-from-mac-os-x-snow-leopard-logon-screen
+            sudo dscl . -delete /Users/_centralreport AuthenticationAuthority
+            sudo dscl . -create /Users/_centralreport Password "*"
+
+            # Registring the _centralreport user in the _centralreport group
+            sudo dscl . -append /Groups/_centralreport GroupMembership _centralreport
+
+            # Checking if the user have been created
```
Comment (Ninir) [inline bash/functions.inc.sh]: Checking that the user has been created

File: bash/functions.inc.sh [hunk 31]
```diff
@@ -1,79 +1,823 @@
...
+
+    # Uninstall existing previous installation, if exist
+    displayAndExec "Stopping CentralReport..." stop_cr
+    RETURN_CODE="$?"
+    if [ ${RETURN_CODE} -ne 0 ]; then
+        return ${RETURN_CODE}
+    fi
+
+    # Delete CR binary script
+    displayAndExec "Removing binary file..." delete_bin
```
Comment (Ninir) [inline bash/functions.inc.sh]: `Removing the binary file...`

File: bash/functions.inc.sh [hunk 32]
```diff
@@ -1,79 +1,823 @@
...
+    RETURN_CODE="$?"
+    if [ ${RETURN_CODE} -ne 0 ]; then
+        return ${RETURN_CODE}
+    fi
+
+    # Delete CR lib files
+    displayAndExec "Removing CentralReport library..." delete_lib
```
Comment (Ninir) [inline bash/functions.inc.sh]: `Removing the CentralReport library...`

or

`Removing CentralReport libraries...`

File: bash/functions.inc.sh [hunk 33]
```diff
@@ -1,79 +1,823 @@
...
+    RETURN_CODE="$?"
+    if [ ${RETURN_CODE} -ne 0 ]; then
+        return ${RETURN_CODE}
+    fi
+
+    # Delete startup plist file
+    displayAndExec "Removing startup plist..." delete_init_file
```
Comment (Ninir) [inline bash/functions.inc.sh]: `Removing the CentralReport startup plist file...`

File: bash/functions.inc.sh [hunk 34]
```diff
@@ -1,79 +1,823 @@
...
+    RETURN_CODE="$?"
+    if [ ${RETURN_CODE} -ne 0 ]; then
+        return ${RETURN_CODE}
+    fi
+
+    printTitle "Installing CentralReport..."
+
+    displayAndExec "Creating system user..." create_cr_user
```
Comment (Ninir) [inline bash/functions.inc.sh]: `Creating the system user...`

File: bash/functions.inc.sh [hunk 35]
```diff
@@ -1,79 +1,823 @@
...
+    RETURN_CODE="$?"
+    if [ ${RETURN_CODE} -ne 0 ]; then
+        return ${RETURN_CODE}
+    fi
+
+    displayAndExec "Copying CentralReport binary file..." copy_bin
```
Comment (Ninir) [inline bash/functions.inc.sh]: `Copying the CentralReport binary file...`

File: bash/functions.inc.sh [hunk 36]
```diff
@@ -1,79 +1,823 @@
...
+    RETURN_CODE="$?"
+    if [ ${RETURN_CODE} -ne 0 ]; then
+        return ${RETURN_CODE}
+    fi
+
+    displayAndExec "Copying CentralReport library..." copy_lib
```
Comment (Ninir) [inline bash/functions.inc.sh]: `Copying the CentralReport library...`

File: bash/functions.inc.sh [hunk 37]
```diff
@@ -1,79 +1,823 @@
...
+    RETURN_CODE="$?"
+    if [ ${RETURN_CODE} -ne 0 ]; then
+        return ${RETURN_CODE}
+    fi
+
+    displayAndExec "Creating configuration directory..." create_config_directory
```
Comment (Ninir) [inline bash/functions.inc.sh]: `Creating the CentralReport configuration directory...`

File: bash/functions.inc.sh [hunk 38]
```diff
@@ -1,79 +1,823 @@
...
+    RETURN_CODE="$?"
+    if [ ${RETURN_CODE} -ne 0 ]; then
+        return ${RETURN_CODE}
+    fi
+
+    displayAndExec "Creating log directory..." create_log_directory
```
Comment (Ninir) [inline bash/functions.inc.sh]: `Creating the CentralReport log directory...`

File: bash/functions.inc.sh [hunk 39]
```diff
@@ -1,79 +1,823 @@
...
+    RETURN_CODE="$?"
+    if [ ${RETURN_CODE} -ne 0 ]; then
+        return ${RETURN_CODE}
+    fi
+
+    displayAndExec "Copying CentralReport init file..." copy_init_file
```
Comment (Ninir) [inline bash/functions.inc.sh]: `Creating the CentralReport init file...`

File: bash/functions.inc.sh [hunk 40]
```diff
@@ -1,79 +1,823 @@
...
+        return ${RETURN_CODE}
+    fi
+
+    # Cleaning screen
+    clear
+
+    # CR config assistant
+    # TODO: Refactor config assistant
+
+    printTitle "First start of CentralReport..."
```
Comment (Ninir) [inline bash/functions.inc.sh]: To see: `First CentralReport launch...`

File: bash/functions.inc.sh [hunk 41]
```diff
@@ -1,79 +1,823 @@
...
+
+    # Check if CentralReport is already running, and stop it.
+    displayAndExec "Stopping CentralReport..." stop_cr
+    RETURN_CODE="$?"
+    if [ ${RETURN_CODE} -ne 0 ]; then
+        return ${RETURN_CODE}
+    fi
+
+    # Delete startup plist file
+    displayAndExec "Removing CentralReport startup plist..." delete_init_file
```
Comment (Ninir) [inline bash/functions.inc.sh]: `Removing the CentralReport startup plist...`

File: bash/functions.inc.sh [hunk 42]
```diff
@@ -1,79 +1,823 @@
...
+    RETURN_CODE="$?"
+    if [ ${RETURN_CODE} -ne 0 ]; then
+        return ${RETURN_CODE}
+    fi
+
+    # Delete CR bin files
+    displayAndExec "Removing CentralReport binary script..." delete_bin
```
Comment (Ninir) [inline bash/functions.inc.sh]: `Removing the CentralReport binary script...`

File: bash/functions.inc.sh [hunk 43]
```diff
@@ -1,79 +1,823 @@
...
+    RETURN_CODE="$?"
+    if [ ${RETURN_CODE} -ne 0 ]; then
+        return ${RETURN_CODE}
+    fi
+
+    # Delete CR lib files
+    displayAndExec "Removing CentralReport library..." delete_lib
```
Comment (Ninir) [inline bash/functions.inc.sh]: `Removing the CentralReport library...`

File: bash/functions.inc.sh (hunk 22 above)
Comment (miniche) [inline bash/functions.inc.sh]: One user for CentralReport, it's enough :)

**Comment by Ninir** [comment]
Euh... oui oui, that's all good... eh...

**Comment by miniche** [comment]
Ready to merge? :neckbeard:

**Comment by Guimove** [comment]
Yep

**Pull request merged**